### PR TITLE
Implement for Haiku 

### PR DESCRIPTION
**Pull request by Kijewski**
Tested on Haiku R1/beta3.

The stress-test takes 1m43s. That's quite terrible.

### CODE REVIEW

**Comment by lopopolo** [comment]
Is binding to the haiku C++ API with a small native extension compiled by `cc` in `build.rs` an option here? This looks like it would be a lot simpler and less error prone since it is one API call.

https://www.haiku-os.org/docs/api/classBTimeZone.html

**Comment by Kijewski** [comment]
Sure! Do you have experience how to work with `cxx`?

[`BLocaleRoster::GetDefaultTimeZone()`](https://www.haiku-os.org/docs/api/classBLocaleRoster.html#a7cb5f911107993a9c4816b52e9221572) + [`BTimeZone::ID()`](https://www.haiku-os.org/docs/api/classBTimeZone.html#afa38541edc05b13c25618f1bca75c6b0) + [`BString::CopyInto(&buf, 0, i32::MAX)`](https://github.com/haiku/haiku/blob/530f89aa6d0a551aa8ab4a3af16d88186f04878c/src/kits/support/String.cpp#L471) should do exactly what we need.

**Comment by lopopolo** [comment]
I don't have experience with ccx to call this directly in Rust, but a tiny `.cpp` file with a C ABI export and one header can be made pretty usable with the `cc` crate and bindgen in a `build.rs` script

**Comment by Kijewski** [comment]
It works. The stress-test now executes in 19 seconds.

Well, at least on an x86_64 system it works. But even if it only works for 64bit systems, I guess that's good enough.

**Comment by lopopolo** [comment]
This is looking great!

If you'd like to add clang-format to CI, we can use: https://github.com/artichoke/clang-format. I use it in Artichoke for C code. I recently added C++ support to the file scanner.

I tried it out with a simple `.clang-format` config and came up with this:

```diff
commit 20fa465a054534b98955a721ad5ba7dad03e851a (HEAD -> pr-haiku)
Author: Ryan Lopopolo <rjl@hyperbo.la>
Date:   Wed Sep 28 19:41:25 2022 -0700

    Add clang-format config

diff --git a/.clang-format b/.clang-format
new file mode 100644
index 0000000..444c17f
--- /dev/null
+++ b/.clang-format
@@ -0,0 +1,9 @@
+---
+BasedOnStyle: Google
+IndentWidth: 4
+ColumnLimit: 100
+---
+Language: Cpp
+SortIncludes: false
+TabWidth: 4
+UseTab: Never
diff --git a/include/impl_haiku.h b/include/impl_haiku.h
index cd50fe8..2d9c362 100644
--- a/include/impl_haiku.h
+++ b/include/impl_haiku.h
@@ -5,5 +5,5 @@
 #include <cstddef>

 namespace tz_haiku {
-    size_t get_tz(uint8_t *buf, size_t buf_len);
+size_t get_tz(uint8_t *buf, size_t buf_len);
 }
```

A GitHub Actions workflow to check this in CI looks like:

```yaml
  c:
    name: Lint and format C
    runs-on: ubuntu-latest
    steps:
      - name: Checkout repository
        uses: actions/checkout@v3

      - name: Setup Node.js runtime
        uses: actions/setup-node@v3
        with:
          node-version: 16

      - name: Install npm
        run: npm i -f -g npm@8.16.0

      - name: Lint and check formatting with clang-format
        run: npx github:artichoke/clang-format --check
```

File: src/tz_haiku.rs
Comment (lopopolo) [inline src/tz_haiku.rs:22]: I think this reads a bit clearer and avoids doing the length check _and_ a panicking index operation.

```suggestion
    match buf.get(..len) {
        Some(s) if s.is_empty() => Err(crate::GetTimezoneError::OsError),
        Some(s) => {
            let s = std::str::from_utf8(s).map_err(|_| crate::GetTimezoneError::OsError)?;
            Ok(s.to_owned())
        }
        None => Err(crate::GetTimezoneError::OsError),
    }
```

**Review by lopopolo — CHANGES_REQUESTED** [review]
I left a suggested change to make the Rust side of things a bit easier to read and I was hoping we could add some comments in a couple of places on the C++ side of things.

Thanks for digging into cxx and the Haiku C++ docs. I like this a lot better than the initial pass at this PR.

File: src/tz_haiku.rs
Comment (Kijewski) [inline src/tz_haiku.rs:7]: This function can be implemented much simpler without using `unsafe` or casting slices to pointers manually: https://cxx.rs/binding/slice.html. I did not notice this feature yesterday. Will change it this evening or tomorrow.

File: src/impl_haiku.cc
Comment (Kijewski) [reply src/impl_haiku.cc:33]: Yes, the string is borrowed. I'll add a comment.

> You can't modify or free the pointer. Once the [BString](https://www.haiku-os.org/docs/api/classBString.html) object is deleted, the pointer becomes invalid.

https://www.haiku-os.org/docs/api/classBString.html#ae4fe78b06c8e3310093b80305e14ba87  
https://github.com/haiku/haiku/blob/530f89aa6d0a551aa8ab4a3af16d88186f04878c/headers/os/support/String.h#L464-L470

File: src/impl_haiku.cc
Comment (Kijewski) [reply src/impl_haiku.cc:10]: Haiku only declares their `malloc` / `realloc` / `free` / `operator new` functions as `noexcept`. On amd64's ABI a `try` block should have almost no overhead, if there was no exception thrown. Putting everything into a `try` block seems like a good idea.

**Comment by Kijewski** [comment]
I don't understand how this change could break the wasm implementation?? https://github.com/strawlab/iana-time-zone/actions/runs/3148078221/jobs/5118236934

File: src/impl_haiku.cc
Comment (lopopolo) [reply src/impl_haiku.cc]: thanks for adding this.

File: include/impl_haiku.h
Comment (lopopolo) [reply include/impl_haiku.h]: this was addressed by using `rust::Slice` type from `cxx`.

File: src/tz_haiku.rs
Comment (lopopolo) [reply src/tz_haiku.rs]: done in https://github.com/strawlab/iana-time-zone/pull/66/commits/742f223624a46326a7e4d304d4894c6c38950742

**Comment by lopopolo** [comment]
I'm not sure either, but I bumped the minimum version of `js-sys` which seems to have fixed it for me locally

File: Cargo.toml
Comment (lopopolo) [inline Cargo.toml:27]: if I had to take a guess at why this is required, `cxx` probably has a different minimum version of `syn` or some other proc macro crate which doesn't play nice with earlier versions of `js-sys`.

File: src/impl_haiku.cc
Comment (lopopolo) [inline src/impl_haiku.cc:42]: we're relying on integer promotion rules here and also cast `length` to `size_t` in several places.

File: src/impl_haiku.cc
Comment (lopopolo) [inline src/impl_haiku.cc:29]: I don't know if this actually compiles, but can we do something like this:

```suggestion
        size_t length(size_t(bname.Length()));
```

**Review by lopopolo — CHANGES_REQUESTED** [review]

**Comment by lopopolo** [comment]
This is looking good! One more question on the C++ side.

File: src/impl_haiku.cc
Comment (lopopolo) [reply src/impl_haiku.cc:29]: ahh maybe it's not that simple since we check if `length < 0` just beyond this. This is a spot where I prefer Rust shadowing ability.

File: Cargo.toml
Comment (Kijewski) [reply Cargo.toml:27]: Thanks. Yes, something like that seems likely.

File: src/impl_haiku.cc
Comment (Kijewski) [reply src/impl_haiku.cc]: I introduced a new variable after checking that it's not negative. Yes, rust's variable shadowing is a very nice feature that I guess would work well with C++'s scoping rules, too. But I don't think that C++ will ever adopt this feature. Otherwise code like this wouldn't compile anymore: https://godbolt.org/z/8s8zM9K6h

```cpp
int main(int, char**) {
    int result(result);
    return result;
}
```

**Review by lopopolo — APPROVED** [review]

**Comment by lopopolo** [comment]
@astraw this looks good to me. Do you want to take a quick look before we start the 7 day clock on this unsafe-proposed PR?

**Comment by astraw** [comment]
Thanks.

However, I think a `build.rs` on all platforms is a heavy price to pay to support a single platform, namely haiku. Can we move the haiku specific build.rs and cxx stuff to a separate crate (in the same git repo) so that `build.rs` is isolated to that platform?

**Comment by Kijewski** [comment]
build.rs's main function is empty for other OSes. `cxx` is only included for that one OS, because everything is wrapped in

```rust
    #[cfg(target_os = "haiku")]
    let _: () = {
        ....
    }
```


I don't think that having an empty build script would make a noticeably difference.

**Comment by astraw** [comment]
The concern is that the build script must be executed (not what it executes per se). I'm not so much worried about execution time etc but rather about cross compilation or using other toolchains etc.

**Comment by Kijewski** [comment]
Cross compilation is not affected, because the build script is always only executed on the host. At least I don't see how it could be affected right now.

E.g. if your build.rs to read `/etc/hostname` that will work on a linux host even when cross compiling to windows. But it will fail on a windows host, even when cross compiling to linux.

**Comment by astraw** [comment]
Yes, a `build.rs` file triggers building and execution of code on the host during cross compilation. My point is that this brings  complexity for no gain on all platforms except Haiku. To me, the alternative is relatively painless and only paid by people building for Haiku.

**Comment by Kijewski** [comment]
Yeah, also this way users of other OS don't need to wonder why iana-time-zone executes a build script all of a sudden. I split off the code into its own crate.

File: haiku/src/implementation.cc
Comment (astraw) [inline haiku/src/implementation.cc:21]: Can you add a comment here saying that `BLocaleRoster::Default()` returns a static pointer and thus no free is necessary?

**Review by astraw — CHANGES_REQUESTED** [review]
Apart from the one trivial change, this looks good to me.

File: haiku/src/implementation.cc
Comment (Kijewski) [reply haiku/src/implementation.cc:23]: The [documentation](https://www.haiku-os.org/docs/api/classBLocaleRoster.html#a7cb5f911107993a9c4816b52e9221572) is not terribly explicit about what to do with the returned pointer, but at the place of the implementation [a comment](https://github.com/haiku/haiku/blob/8f16317a5b6db5c672f331814273e5857555020f/src/kits/locale/LocaleRoster.cpp#L143-L147) indicates that the pointer is static. → [`MutableLocaleRoster::Default()`](https://github.com/haiku/haiku/blob/8f16317a5b6db5c672f331814273e5857555020f/src/kits/locale/MutableLocaleRoster.cpp#L59-L66)

**Review by astraw — APPROVED** [review]

**Comment by astraw** [comment]
We have 2 approvals now, so according to our `Unsafe-Proposed` policy (#64), let's let this sit until 8 October and then merge.

**Comment by astraw** [comment]
@Kijewski thanks for the work on this, including shifting most of it to its own crate.

**Pull request merged**
